### ISSUE TITLE
Calling functions in a module without any use

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#13523]
Should all top-level modules be available even without being `use`d? As with `SomeModule.someFunction()` when there is no `use SomeModule`?

### Very Simple Example

``` chapel
// A.chpl
module A {
  writeln("in A");
  // Note: no use B
  B.bfunc();
}
```

``` chapel
// B.chpl
module B {
  proc bfunc() {
    writeln("in B.bfunc");
  }
}
```

This does not compile today if you run simply `chpl A.chpl`. (It is possible to compile it by adding `B.chpl` to the compile line, but that is considered to add a `use`, and so the question remains).

### Application Example
``` chapel
// Program.chpl
module Program {
  writeln("in Program");
  // Note: no use B
  MatrixLibrary.matrixFunc();
  use HashingLibrary only;
  HashingLibrary.hashingFunc();
}
```

``` chapel
// MatrixLibrary.chpl
module MatrixLibrary {
  proc matrixFunc() {
    writeln("in MatrixLibrary.matrixFunc");
  }
}
```

``` chapel
// HashingLibrary.chpl
module HashingLibrary {
  private use MatrixLibrary; // for internal use only... try removing it
  proc hashingFunc() {
    writeln("in HashingFunc.hashingFunc");
  }
}
```

This program happens to compile today with `chpl Program.chpl`. However, if the `private use MatrixLibrary` in `HashingLibrary.chpl` is removed, it does not compile. That seems surprising and undesirable since now unrelated changes in one library affects whether on not the main program will compile.

### Possible Directions

I know of two possible directions to go to resolve this issue:
 1. Do not allow code like `SomeModule.someFunction()` to compile if `SomeModule` is not `use`d (or `import`ed a-la #13119) somewhere in scope.
 2. Allow all calls like `SomeModule.someFunction()` to go looking for a module named `SomeModule` in filesystem according to the module search path and parse it etc.

**Comment by mppf** [comment]
Ultimately, I think that the purpose of modules and `use` statements basically boils down to managing scopes. When we think about `use M only;` or use statements with except lists or the proposal for `import`, we are considering how users can have explicit control of the symbols available in their code.

I think that it's important that users can manage which top-level module names are visible as regular variables within their program.  In the future, optimistically, there will be a large number of top-level modules available (presumably as mason packages). Some will undoubtedly have poorly chosen names. Simply expecting mason to handle this problem is not good enough since it runs into the same kinds of issues as Application Example above.

What if, for example, a library has a type with the same name as top-level module? Wouldn't we want the library author to be able to control whether or not that top-level module name refers to the module or to the function?

Suppose there is a module called `Vector` and they want a type of their own called `Vector` and *also* to use the module. They can do something like `use Vector as VectorModule;` to control the name conflict under option 1, but this will not resolve the problem under option 2, since `Vector.someFunction()` will always refer to the top-level module. (Yes, there might be another way that they can do this with shadowing, but that also has pitfalls, such as if trying to `use` two different modules, where one defines a type `Vector` and the other one is a module named `Vector`).

Historically, there was some expectation that modules could be discovered even when the filename does not match the module name (see https://github.com/chapel-lang/chapel/issues/12923#issuecomment-489302742 ). I do not think we should do that, because I view it as a feature that the source code to a module can be easily found by a human. However, if we did, it might be more surprising to have `SomeModule.someFunction()` go and find `someFunction()` within `SomeOtherFilename.chpl`.

As an implementation matter, there is a bit of a chicken-and-egg problem here with the names of the top-level modules themselves. If top-level module names are not always available, what does `use M` actually do? `use` and `import` could continue to work if the module name is already available in the current scope (as I would expect to be the case with submodules or with modules already use'd or used and renamed; or with `use` statements that exist to re-export or make visible names within a module). However, if the module name doesn't already refer to a module, it would go looking in the module search path for a module with that name and then after such a `use M`, `M` would be in scope to refer to that module.

If there were no `use SomeModule`, and we had `SomeModule.someFunction()`, this strategy would lead to a compilation error, because the symbol `SomeModule` in `SomeModule.someFunction()` would not refer to anything yet. (Even though it is the name of a module sitting somewhere on the filesystem... only `use` can use that way of finding it).

Lastly, I can think of plenty of languages that require some sort of `import` to access a module, but I don't know of any that allow a call like `SomeModule.someFunction()` to find the module in the filesystem. For a scripting language, this kind of thing would be particularly surprising, since the process of loading the module can run arbitrary module initialization code. I think as a result scripting languages tend to keep it to relatively few constructs like `import`.

**Comment by lydia-duncan** [comment]
In the language as it currently stands, I think it is expected that qualified accesses (`SomeModule.someFunction()`) are acceptable so long as the module can be found in the program itself.

I do dislike that this behavior varies rather than being consistent, but would prefer to see it change in favor of finding the module for the qualified access rather than requiring `use SomeModule only;` or `use SomeModule except *;` more frequently.  That said, when `import` statements are added to the language, I would be okay with us requiring them for qualified accesses.

**Comment by ben-albrecht** [comment]
I feel strongly that we should pursue option 1.

Not having the list of modules used in the source code impacts productivity negatively IMO. It could be argued that tooling could eliminate this issue by detecting and listing the used modules. However, such tooling does not exist for Chapel today and probably won't for some time. Additionally, modern tooling in other languages achieves the same effect by managing module usage automatically.

**Comment by mppf** [comment]
See also  #13536 which asks the same question for submodules.

**Comment by bradcray** [comment]
Thinking out loud:  Is the following an option?

3) Starting from the motivating example here, it seems to me that since the `use` of `MatrixLibrary` is private in the hashing library, it shouldn't make the `MatrixLibrary` symbol available for other modules that `use` it to see.  I.e., only a public `use`s would result in making a module name available/known at global / program scope (as all `use`s seem to today).

I find this preferable to a strict interpretation of option 1 because it seems that option 1 would prevent me from being able to extend a module or bundle a group of modules together in a new module via `public use` such that anyone who used my module would have access to the symbols from the wrapped module(s) that I wanted them to.  E.g., `ChapelStandard.chpl` is an example of this where we rely on `use` to make a bunch of standard modules automatically available without requiring each to be used in the programmer's scope.  I think similar bundlings and specializations would be useful to programmers as well.

OTOH, maybe the above is wrong and that there's a difference between the `use ChapelIO` in `ChapelStandard` making things like `ChapelIO`'s symbols available (e.g., `writeln()`) and making the symbol `ChapelIO` itself available.  I.e., By having ChapelStandard `use ChapelIO` to expose `writeln()`, my goal is presumably to make `writeln()` available, not necessarily the `ChapelIO` symbol itself.

So maybe that suggests I'm really behind option 1.

I'm not a fan of option 2.  It seems way too fast and loose and heavyweight to have any typo (e.g., `myCC.foo();` where I meant to type `myC.foo();` and `myC` is a class in the scope, say) go and look for a module just because I mistyped.  Of course I could mistype in `use MM;` as well and get the wrong module, but there's no way to guard against all typos.

It'd be interesting to see what the impact of option 1 would have on existing code.

**Comment by mppf** [comment]
> By having ChapelStandard use ChapelIO to expose writeln(), my goal is presumably to make writeln() available, not necessarily the ChapelIO symbol itself.

I think it's important to think about the visibility of the module symbols themselves (`ChapelIO` in this case). I think that we should make `use` and `import` have different behavior in this regard - that `public import ChapelIO;` would make `ChapelIO.foo()` available in code using ChapelStandard; but just `use ChapelIO;` would leave `ChapelIO` itself private (while making all the functions, like `halt`, visible in code using ChapelStandard). (see also  https://github.com/chapel-lang/chapel/issues/13119#issuecomment-520984841 )

**Comment by mppf** [comment]
Here is another form of this problem

> ``` chapel
> // in M.chpl
> module L {
>   proc lFunction() { }
> }
> module N {
>   proc main() {
>     L.lFunction();  // why does this work?
>   }
> }
> ```


This came from https://github.com/chapel-lang/chapel/issues/13536#issuecomment-514673306
and https://github.com/chapel-lang/chapel/issues/13536#issuecomment-521053538 in which @bradcray pointed out

>  When they're in the same file, `L` is known simply by virtue of the fact that you asked the compiler to compile a module named `L` by handing it its source code (and doing so puts it into the global scope as illustrated above, so N can see it, as with any other top-level module).

@bradcray would you agree that if we chose Option 1 here than we'd make the case above no longer work? (It seems to be the same issue to me...) Or would you consider it a separate problem?

**Comment by lydia-duncan** [comment]
I would consider it a separate problem, but could see that being confusing.  If the two must be linked, I think I would be more strongly in favor of option 2, because I care more about the behavior between modules in the same file than I do about the interactions between modules in different files.

As a user, I more readily forgive having to explicitly include more modules on my compilation command than I do additional use statements to use a symbol that, were it not a module, I would otherwise be able to reference (this is kind of a rephrasing of https://github.com/chapel-lang/chapel/issues/13536#issuecomment-521053538)

**Comment by bradcray** [comment]
> > By having ChapelStandard use ChapelIO to expose writeln(), my goal is presumably to make writeln() available, not necessarily the ChapelIO symbol itself.
>
> I think it's important to think about the visibility of the module symbols themselves (ChapelIO in this case). I think that we should make use and import have different behavior in this regard - that public import ChapelIO; would make ChapelIO.foo() available in code using ChapelStandard; but just public use ChapelIO; would leave ChapelIO itself private (while making all the functions, like halt, public and visible in code using ChapelStandard). (see also #13119 (comment) )

Thinking about this piece of the issue for a few more days after writing my last comment (and before reading yours), I've found myself backpedaling a bit...  I'm increasingly thinking that the factor that makes me want to avoid referring to `ChapelIO` has more to do with it being an internal module than it does with the fact that another module `use`s it to propagate it.  I.e., maybe the rule I really wanted was "you just can't refer to internal modules by name (because users of Chapel should never really need to know that they exist and because we want to be able to change their names and refactor that code as much as we like).

Maybe put another way, if I had a project that involved some library modules that were heavily-used:

```chapel
MyProjMath.chpl:
  proc myMathFn() { ... }

MyProjPlot.chpl:
  proc myPlotFn() { ... }

MyProjIO.chpl:
  proc myIOFn() { ... }
```

And I rolled them all up for convenience using one meta-module:

```chapel
MyProj.chpl:
  use MyProjMath, MyProjPlot, MyProjIO;
```

I don't know that I'd want the convenience of being able to type `use MyProjLibs;` to access all three modules to necessarily lock me out of being able to refer to the component functions in a more fully-qualified manner as `MyProjMath.myMathFn()`, `MyProjPlot.myPlotFn()`, and `MyProjIO.myIOFn()`, would I?

So maybe flipping it around, is there an argument for making `use` not make the module name available / visible?

**Comment by bradcray** [comment]
> @bradcray would you agree that if we chose Option 1 here than we'd make the case above no longer work?

Yes, I think choosing option 1 would make the reference to `L` in `N` an error for that code snippet.

**Comment by lydia-duncan** [comment]
> So maybe flipping it around, is there an argument for making use not make the module name available / visible?

I've been going through various notes in prep for proposing an `import` implementation, and found my chart of Haskell vs. Chapel imports - I figured it was a useful data point here.  It looks like Haskell always provides the qualified access for a symbol (`Mod.x`), but gives an option to exclude the unqualified access (e.g. `import Mod(x)` allows `x` and `Mod.x` but `import qualified Mod(x)` only allows `Mod.x`)

**Comment by lydia-duncan** [comment]
Correction: There is also a way to import nothing from a module (neither qualified or unqualified access) using `import Mod()`, but I dismissed that because I thought it was odd

**Comment by mppf** [comment]
Responding to @bradcray in https://github.com/chapel-lang/chapel/issues/13523#issuecomment-522783098

> I'm increasingly thinking that the factor that makes me want to avoid referring to ChapelIO has more to do with it being an internal module than it does with the fact that another module uses it to propagate it.

Right, but users will encounter the same structures and desires. (I.e. even though we have an idea of an "internal" module in the compiler, users writing Mason packages will want to create "internal" submodules as well - that is, submodules that the user of the package cannot refer to - even as they use functions inside of them). 

> I don't know that I'd want the convenience of being able to type use MyProjLibs; to access all three modules to necessarily lock me out of being able to refer to the component functions in a more fully-qualified manner as MyProjMath.myMathFn(), MyProjPlot.myPlotFn(), and MyProjIO.myIOFn(), would I?

I think this is a reasonable criticism of the proposal to make `use` not bring in the symbol. 

> So maybe flipping it around, is there an argument for making use not make the module name available / visible?

I think that there should be *some* way to implement a module with a collection of other (sub-
)modules without exposing the implementation to users. I think that one should be able to choose between this and the other outcome when creating the module.

Let's consider your example and how these two variations could be expressed.

Under the use/import difference proposal, if the module author wanted to expose MyProjMath, MyProjPlot, and MyProjIO as modules, they could `import MyProjMath, MyProjPlot, MyProjIO;` and then `use MyProjMath, MyProjPlot, MyProjIO;` in the meta-module. If they viewed MyProjMath, MyProjPlot, and MyProjIO as "internal" modules aka implementation details, they would just write `use MyProjMath, MyProjPlot, MyProjIO;`. So both behaviors are available.

Now, another option is to follow #13528 (to "re-export" something in a private module). In this setting, we would need to make MyProjMath, MyProjPlot, MyProjIO submodules of MyProj. Making them public modules (and using them) leads to the visible bundling behavior. To make them internal/implementation details, make them private submodules, and then do a `public use` to make their contents public again. 

However since #13528 is just about submodules it is less clear how it could apply to a module bundling other modules. For example, suppose that MyProj is a mason package that uses a GNUPlot mason package. Should MyProj be able to choose whether or not the fact that GNUPlot is implementing the provided plot routines is visible to clients of MyProj? Here I think the case is less strong than in the submodule case. However I was proposing the use/import difference because it could work in this settings as well.

I suppose we could say something like "use does not change the visibility of the module symbol itself if it is already visible". This might be viewed as an extension of #13528. Then, the `use GNUPlot` within MyProj by itself would make `MyProj.GNUPlot` available to users of MyProj. But if MyProj wanted to make the fact that GNUPlot was involved private, the could write

``` chapel
module MyProj {
  private import GNUPlot;
  public use GNUPlot;
}
```

to indicate (more explicitly) that GNUPlot itself is an implementation detail, even though  its contents for part of the public API. In order for that to work, the `public use GNUPlot` would need to see that there `GNUPlot` was already imported and therefore not change the privacy of the `GNUPlot` symbol itself. That way, the GNUPlot symbol remains private.

I like that this later idea makes what is arguably the stranger/less common behavior (hiding the implementation structure) longer/more explicit while keeping `use MyProj` working for the more typical case (for users of MyProj, having MyProj.f() available in the file is fine). However I suspect that @lydia-duncan will have something to say about it.

**Comment by lydia-duncan** [comment]
Those are interesting considerations.  However, I'm not concerned about bleeding module names into namespaces due to `use` statements - I'd much rather we get an error about it as soon as there is a name conflict between modules and other symbols rather than only when `use`/`import` statements occur, as that will lead to less bugs discovered down the road due to insufficient testing.

Of course, that might open a whole other can of worms w.r.t. being a single namespace language :)

**Comment by bradcray** [comment]
I'm behind on this issue again, but as a relevant side-note:  In PR #13822, I'm trying to pull `use` statements out of `ChapelStandard.chpl` that aren't really used to define the language, but have been put there for convenience or laziness.  As an example, `HaltWrappers` doesn't really help define the Chapel language itself and isn't something we've decided to make part of the standard set of `use`d modules (though it was postfixed with an `only ;` to prevent symbols from being available by default), but it was listed in ChapelStandard.chpl.  So I removed it and only added it to modules that required it to get testing passing again.  In reviewing the code, this caused Elliot to balk because it seemed odd that modules invoking `HaltWrappers` routines did not themselves contain `use HaltWrappers` statements (directly or, as far as we could tell, transitively).  And it seems natural that in order to use routines from a given module you should arguably have to say "I want to use routines from this module" (or inherit them from another module's public use).

That said, there's also some strange stuff going on here where unqualified references to HaltWrappers symbols seem to be resolving even though it's not clear how they could be (i.e., I don't think `public use`s of HaltWrappers remain other than locally scoped ones, so don't think the symbols should resolve).  I may need @lydia-duncan's help to debug those once I've double-checked them and possibly distilled them down to a simple example.

Anyway, this exercise has suggested to me that approach 1 may be the correct approach (and certainly doesn't seem wrong).  The thing that remains slightly odd about it to me is that it makes `module M { ... }` declarations fairly different than other declarations of new symbols in that they're not automatically visible if they're found in a parent scope.  But if we think of Michael's proposal that `use M` (or `use M as ...`) is the thing that is declaring the symbol (i.e., it is the sibling to `proc`, `var`, `iter`, etc.) and that module declarations are different in some way, then I think that arguably helps put things back in order.

**Comment by lydia-duncan** [comment]
> That said, there's also some strange stuff going on here where unqualified references to HaltWrappers symbols seem to be resolving even though it's not clear how they could be

That sounds like another instance of #7200 which I only remembered existed due to looking through the history for the `import`

**Comment by bradcray** [comment]
I think it's different.  In this case, the behavior was after `use HaltWrappers only ...;` was removed.  I don't want this question to hijack this issue, though.  I'll open up another when I can describe it better.

**Comment by bradcray** [comment]
> users writing Mason packages will want to create "internal" submodules as well - that is, submodules that the user of the package cannot refer to - even as they use functions inside of them

This seems like it might be good motivation for supporting `private module` declarations with `public use`s of them at the same scope (as we've been discussing on issue #13528)?  I.e., `private module M` means that things outside this scope can't refer to the symbol `M`, but `public use M` would make the (public) contents of `M` visible outside the scope?  Then, once we can store sub-modules in a different file, all of the stuff used by ChapelStandard to implement the language (Atomics, ChapelRange, DefaultRectangular, etc.) could all be private submodules of ChapelStandard?  (which should probably really be renamed to something like `ChapelLanguage`).

(Oh, I see you got to a similar place... that's what I get for answering a paragraph at a time.  But at least it suggests we're on a similar mental page).

For your GNUPlot case, in a "change `use` as little as possible" / "avoid import for now" world, I might suggest that a way to hide the name of a `use`d module might be to support one or both of the following (assuming we add an `as` clause to `use` statements as proposed in #10799):

```chapel
use GnuPlot as _;
use GnuPlot as ;
```

that is, "make GnuPlot's symbols visible, but drop the name `GnuPlot` itself on the floor."

**Comment by mppf** [comment]
Responding to @lydia-duncan  https://github.com/chapel-lang/chapel/issues/13523#issuecomment-523213870

> However, I'm not concerned about bleeding module names into namespaces due to `use` statements... Of course, that might open a whole other can of worms w.r.t. being a single namespace language :)

I'm not really on board with Chapel being a "single namespace language", depending on the definition.

I think the proper definition is what @bradcray posted here https://github.com/chapel-lang/chapel/issues/6698#issuecomment-493269231

> What I think the term has traditionally been intended to be used for w.r.t. Chapel is to say that we don't treat identifiers of one kind differently than identifiers of another kind (i.e., there isn't a collection of function names distinct from the collection of variable names).

I think that the idea that all modules exist in a single, global namespace, is something that we're trying to fix. We're trying to fix it by leaning on the idea of submodules to create the heirarchy, which means that you might view "top-level" modules as forming a single global namespace; however I've been arguing:
 1. `use` and `import` should be the only things interacting with this top-level module namespace (but then they can re-export the module name into the regular single-namespace)
 2. we need to carefully control the availability of module symbols themselves, when they are pulled into the regular namespace by `use` and `import`.

In other words, I don't think it's reasonable to expect that *all modules including submodules* form a single global namespace. As a result I think we need to be careful about leaking module names, because they can and often will be submodules.

**Comment by vasslitvinov** [comment]
The new rule makes sense to me.

**Comment by BryantLam** [comment]
As an aside, this issue had a small discussion about the following idiom to hide the module symbol but still export its items:
```chapel
module MyProj {
  private import GNUPlot;
  public use GNUPlot;
}
```
Rust refers to this idiom as a facade. [Here are some lessons learned](https://aturon.github.io/blog/2017/07/26/revisiting-rusts-modules/#what-is-a-module-system-anyway) to minimize the mismatch between using the facade and the actual module hierarchy.

**Comment by bradcray** [comment]
Here's a funny (but not ha-ha funny) commit message of mine in Sept 2008:

> Here is perhaps more evidence that we should require modules to be
> used before being used to disambiguate member names.

**Comment by bradcray** [comment]
Oh, and interestingly, commit 2887cddfe56f693dfef96a66ec39ba6bdbea641b (circa 2011) indicates that we did switch to that behavior for some time, and then switched back.  It's amazing how much you can forget in 8 years...
